safeio/rules.go

```diff
@@ -0,0 +1,100 @@
+package safeio
+
+import (
+	"strings"
+)
+
+// FSAccessRule knows how to allow or deny actions on a certain file path.
+type FSAccessRule interface {
+	GetFSAccessEffect(path string, requested FSAction) (allowed FSAction, denied FSAction)
+}
+
+// FSAccessRuleset allows grouping rules together.
+type FSAccessRuleset struct {
+	Rules []FSAccessRule
+}
+
+// GetFSAccessEffect returns Deny if any of its rules returns Deny, otherwise
+// returns Allow if any of its rules returns Allow, otherwise returns None.
+func (s FSAccessRuleset) GetFSAccessEffect(path string, requested FSAction) (allowed FSAction, denied FSAction) {
+	for _, r := range s.Rules {
+		a, d := r.GetFSAccessEffect(path, requested)
+		allowed |= a
+		denied |= d
+	}
+	return
+}
+
+// FSAccessRulechain chains rules, i.e. for an action to be allowed it needs to
+// be allowed by all members of the chain.
+type FSAccessRulechain struct {
+	Rules []FSAccessRule
+}
+
+func (r FSAccessRulechain) GetFSAccessEffect(path string, requested FSAction) (allowed FSAction, denied FSAction) {
+	allowed, denied = AllFileActions, AllFileActions
+	for _, r := range r.Rules {
+		a, d := r.GetFSAccessEffect(path, requested)
+		allowed &= a
+		denied |= d
+	}
+	return
+}
+
+type PrefixFSAccessRule struct {
+	Prefix         string
+	AllowedActions FSAction
+	DeniedActions  FSAction
+}
+
+func (r PrefixFSAccessRule) GetFSAccessEffect(path string, actions FSAction) (allowed FSAction, denied FSAction) {
+	if !strings.HasPrefix(path, r.Prefix) {
+		// If the path does start with r.Prefix, there is no effect from this rule.
+		return 0, 0
```

Comment:
is returning 0 for `allowed` when the path doesn't start with the prefix going to be a problem for the `FSAccessRulechain`, since it `&=` s the rules? (I see it might false the whole thing where it shouldn't have any effect, right?, or am I missing something here?)

Also you probably meant `... doesn't ...` in the comment that says `... does start with r.Prefix`